pvlib/iotools/sodapro.py

```diff
@@ -0,0 +1,368 @@
+    # added to the base URL as it contains sub-parameters seperated by
+    # semi-colons, which gets incorrectly formatted by the requests function
+    # if passed using the params argument.
+    res = requests.get(base_url + '?DataInputs=' + data_inputs, params=params,
+                       timeout=timeout)
+
+    # Invalid requests returns an XML error message and the HTTP staus code 200
+    # as if the request was successful. Therefore, errors cannot be handled
+    # automatic (e.g. res.raise_for_status()) and errors are handled manually
+    if res.headers['Content-Type'] == 'application/xml':
+        errors = res.text.split('ows:ExceptionText')[1][1:-2]
+        raise requests.HTTPError(errors, response=res)
+    # Successful requests returns a csv data file
+    elif res.headers['Content-Type'] == 'application/csv':
+        fbuf = io.StringIO(res.content.decode('utf-8'))
+        data, meta = parse_cams(fbuf, integrated=integrated, label=label,
+                                map_variables=map_variables)
+        return data, meta
+
+
+def parse_cams(fbuf, integrated=False, label=None, map_variables=True):
+    """
+    Parse a file-like buffer with data in the format of a CAMS Radiation or
+    McClear file. The CAMS services are described in [1]_ and [2]_.
+
+    Parameters
+    ----------
+    fbuf: file-like object
+        File-like object containing data to read.
+    integrated: boolean, default False
+        Whether to return radiation parameters as integrated values (Wh/m^2)
+        or as average irradiance values (W/m^2) (pvlib preferred units)
+    label: {'right', 'left'}, default: None
+        Which bin edge label to label time-step with. The default is 'left' for
+        all time steps except for '1M' which has a default of 'right'.
+    map_variables: bool, default: True
+        When true, renames columns of the Dataframe to pvlib variable names
+        where applicable. See variable CAMS_VARIABLE_MAP.
+
+    Returns
+    -------
+    data: pandas.DataFrame
+        Timeseries data from CAMS Radiation or McClear
+    meta: dict
+        Metadata avaiable in the file.
+
+    See Also
+    --------
+    pvlib.iotools.read_cams, pvlib.iotools.get_cams
+
+    References
+    ----------
+    .. [1] `CAMS Radiation Service Info
+       <http://www.soda-pro.com/web-services/radiation/cams-radiation-service/info>`_
+    .. [2] `CAMS McClear Service Info
+       <http://www.soda-pro.com/web-services/radiation/cams-mcclear/info>`_
+    """
+    meta = {}
+    # Initial lines starting with # contain meta-data
+    while True:
+        line = fbuf.readline().rstrip('\n')
+        if line.startswith('# Observation period'):
+            # The last line of the meta-data section contains the column names
+            names = line.lstrip('# ').split(';')
+            break  # End of meta-data section has been reached
+        elif ': ' in line:
+            meta[line.split(': ')[0].lstrip('# ')] = line.split(': ')[1]
```

Comment:
Not a blocker for me, but just pointing out that this assumes that metadata values are single lines, which isn't true for the `Content` information.  I'm in favor of leaving this as-is instead of building a parser that can handle multiple lines.